autopilot/agent.go

```diff
@@ -342,6 +346,38 @@ func mergeChanState(pendingChans map[NodeID]Channel,
 	return totalChans
 }
 
+// weightedChoice draws a random index from the map of channel candidates, with
+// a probability propotional to their score.
+func weightedChoice(s map[NodeID]ChanCandidate) (NodeID, error) {
+	// Calculate the sum of scores found in the map.
+	var sum float64
+	for _, v := range s {
+		sum += v.Score
+	}
+
+	if sum <= 0 {
+		return NodeID{}, fmt.Errorf("non-positive sum")
+	}
+
+	// Create a map of normalized scores such, that they sum to 1.0.
+	norm := make(map[NodeID]float64)
+	for k, v := range s {
+		norm[k] = v.Score / sum
+	}
+
+	// Pick a random number in the range [0.0, 1.0), and iterate the map
+	// until the number goes below 0. This means that each index is picked
+	// with a probablity eqaual to their normalized score.
+	r := rand.Float64()
+	for k, v := range norm {
```

Comment:
I find this selection a bit hard to analyze. First, the map iteration of `norm` is already randomized. So the probability that a node is selected, is dependent on both the order of map iteration _and_ the random value selected. As an example, if the iteration _wasn't_ randomized, and the value .50 was selected each time, and both nodes had the _same_ score (let's say .50 for the sake of the example), then the first node would _always_ be selected. 

One thing we could do, in order to assert guarantees w.r.t the selection here, is to write a set of `testing/quick` tests that target selection probabilities based on scores+ranges.

Reply:
Added some explanation. Will add some unit/quick tests when the design is settled.

Reply:
Much clearer now with the comment, thanks!